src/solvers/refinement/string_refinement.cpp

```diff
@@ -1716,9 +1744,186 @@ bool string_refinementt::is_axiom_sat(
     }
   case decision_proceduret::resultt::D_UNSATISFIABLE:
     return false;
-  default:
+  case decision_proceduret::resultt::D_ERROR:
     INVARIANT(false, string_refinement_invariantt("failure in checking axiom"));
     // To tell the compiler that the previous line bails
     throw 0;
   }
 }
+
+/// \related string_constraintt
+typedef std::map<exprt, std::vector<exprt>> array_index_mapt;
+
+/// \related string_constraintt
+class gather_indices_visitort: public const_expr_visitort
+{
+public:
+  array_index_mapt indices;
+
+  gather_indices_visitort(): indices() {}
+
+  void operator()(const exprt &expr) override
+  {
+    if(expr.id()==ID_index)
+    {
+      const index_exprt index=to_index_expr(expr);
+      const exprt s(index.array());
+      const exprt i(index.index());
+      indices[s].push_back(i);
+    }
+  }
+};
+
+/// \related string_constraintt
+static array_index_mapt gather_indices(const exprt &expr)
+{
+  gather_indices_visitort v;
+  expr.visit(v);
+  return v.indices;
+}
+
+/// \related string_constraintt
+class is_linear_visitort: public const_expr_visitort
+{
+public:
+  bool correct;
+
+  is_linear_visitort(): correct(true) {}
+
+  void operator()(const exprt &expr) override
+  {
+    if(expr.id()!=ID_plus && expr.id()!=ID_minus && expr.id()!=ID_unary_minus)
+    {
+      // This represents that the expr is a valid leaf, may not be future proof
+      // or 100% enforced, but is correct prescriptively. All non-sum exprs must
+      // be leaves.
+      correct&=(expr.operands().size()==0);
+    }
+  }
+};
+
+/// \related string_constraintt
+static bool is_linear(const exprt &expr)
+{
+  is_linear_visitort v;
+  expr.visit(v);
+  return v.correct;
+}
+
+/// Checks the data invariant for \link string_constraintt
+/// \related string_constraintt
+/// \param [in] expr: the string constraint to check
+/// \return whether the constraint satisfies the invariant
+bool string_refinementt::is_valid_string_constraint(
+  const string_constraintt &expr)
+{
+  // Condition 1: The premise cannot contain any string indices
+  const array_index_mapt premise_indices=gather_indices(expr.premise());
+  if(!premise_indices.empty())
+  {
+    std::string print;
+    for(const auto &pair : premise_indices)
+    {
+      print.append(from_expr(ns, "", pair.first)+": {");
+      for(const auto &i : pair.second)
+        print.append(from_expr(ns, "", i) + ", ");
+      print.append("}");
+    }
+    error() << "Premise has indices: " << from_expr(ns, "", expr.premise())
+            << " => " << from_expr(ns, "", expr.body()) << ", map: {"
+            << print << "}" << eom;
```

Comment:
if `print` is only appended things and output here, you can directly write to error instead of appending